distributed/shuffle/tests/utils.py

```diff
@@ -43,3 +47,54 @@ async def shuffle_barrier(self, id: ShuffleId, run_id: int) -> dict[str, None]:
         for addr, s in self.shuffles.items():
             out[addr] = await s.inputs_done()
         return out
+
+
+class ShuffleAnnotationChaosPlugin(SchedulerPlugin):
+    #: Rate at which the plugin randomly drops shuffle annotations
+    rate: float
+    scheduler: Scheduler | None
+    seen: set
+
+    def __init__(self, rate: float):
+        self.rate = rate
+        self.scheduler = None
+        self.seen = set()
+
+    async def start(self, scheduler: Scheduler) -> None:
+        self.scheduler = scheduler
+
+    def transition(
+        self,
+        key: str,
+        start: TaskStateState,
+        finish: TaskStateState,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        assert self.scheduler
+        if finish != "waiting":
+            return
+        if not key.startswith("shuffle-barrier-"):
+            return
+        if key in self.seen:
+            return
+
+        self.seen.add(key)
+
+        barrier = self.scheduler.tasks[key]
+
+        if self._flip():
+            barrier.annotations.pop("shuffle", None)
+        for dt in barrier.dependents:
+            if self._flip():
+                dt.annotations.pop("shuffle", None)
+
+    def _flip(self) -> bool:
+        return random.random() < self.rate
```

Comment:
Generally, I'm also not a huge fan of this, but I think it is the most defensive way of testing that we do not rely on annotations. Let me disentangle your statement.

> This is testing properties of the current implementation that I consider mostly accidental / nice to haves but not hard requirements.

This is true, but it is also the implementation of annotations that is broken. If it were not broken, we would not need to fix this.

> For instance, I think it is a fair assumption to say that either all tasks or no tasks are annotated. This is generally how fusion and our graph manipulation works. The fact that the implementation can handle a mixture is nice but not required.

At the moment, I do not feel confident to make this claim. I suppose it's true, but I don't know what's broken with annotations at the moment. Also, this feels like an implementation-specific assumption. I could see fusion dropping annotations just for fused tasks if one does not pay attention. I would like to not have to test this, but here we are.

> I think it's also reasonably to say that annotations might be read during runtime and not just initially. I don't know when or how this is useful but I don't see why we should restrict ourselves to the current behavior.

I don't see where we restrict ourselves here. We are stripping the barrier and its dependents of annotations once they arrive on the scheduler's state machine. This makes no assumption about when they will be read. The only assumption this makes is that all the ways that annotations are broken will strip them before the tasks arrive at the scheduler. I think this is a fair assumption.

> I wouldn't want future development to be slowed down if they break these properties / because these tests fail.

Agreed.

Reply:
> At the moment, I do not feel confident to make this claim

Fair enough. It's true that it's hard to extrapolate what happens in the future. 

> I could see fusion dropping annotations just for fused tasks if one does not pay attention.

Indeed. My point is rather that if fusion happens this should affect all output keys. However, this is also a generalization that may not be true. 

> > I think it's also reasonably to say that annotations might be read during runtime and not just initially. I don't know when or how this is useful but I don't see why we should restrict ourselves to the current behavior.

> I don't see where we restrict ourselves here.

Sorry, I see how my statement is not very clear. This thing mutates annotations as soon as the barrier is transitioned to a specific state. If we read out the annotations before this already, this modification will not have *any* effect.
If I'm not mistaken, this is also what's happening in the current implementation. Once the first transfer task is running we're evaluating annotations and are fixing the mappings. This transition hook is only executed later, i.e. the mutated annotations will not have an effect on any of this.
If we later changed the behavior such that the annotations are evaluated later or multiple times we are suddenly hitting this test code which may no longer make sense.

This is all a bit academic. I will not block merging because of this.

> > I wouldn't want future development to be slowed down if they break these properties / because these tests fail.
> 
> Agreed.

What I meant saying is that red tests often discourage people to change something and once this becomes legacy code, the new developer generation may not fully understand the context about whether this is required or just nice-to-have behavior.
I raised my concerns. If future developers run into this, they can follow the breadcrumbs to this thread =)